### DISCUSSION TITLE
Modeling Database architecture [UML]

### DISCUSSION THREAD

**Discussion by bellaabdelouahab** [ISIL-ESTE/Student-Workflow-Organizer#21] [General]
in this discussion we will starting modeling of project depending of our use case needs 
every update about the classes we create or modify will be displayed here ,
feel free to ask question or give suggestions

**Comment by Khalid1G** [comment]
Salam everyone, I would like to add this just like a base UML class diagram: 

> These are some of the main entities that could be included in the UML class diagram for the student organizer website. Depending on the specific requirements and functionalities of the website, we may need to include additional entities or modify these entities accordingly.

```puml
@startuml
class User {
  - username: String
  - password: String
  - email: String
  - preferences: Preferences
}

class Course {
  - courseCode: String
  - name: String
  - description: String
  - resources: List<Resource>
}

class Assignment {
  - title: String
  - description: String
  - deadline: Date
  - status: String
}

class Reminder {
  - title: String
  - description: String
  - date: Date
}

class Collaboration {
  - groupMembers: List<User>
  - discussions: List<Discussion>
  - sharedResources: List<Resource>
}

class ResourceLibrary {
  - resources: List<Resource>
}

class ProgressTracking {
  - completedAssignments: List<Assignment>
  - grades: List<Grade>
  - milestones: List<Milestone>
}

class Analytics {
  - performanceData: PerformanceData
}

class Preferences {
  - theme: String
  - notifications: Boolean
}

class Resource {
  - title: String
  - url: String
}

class Event {
  - title: String
  - description: String
  - date: Date
}

class Discussion {
  - participants: List<User>
  - messages: List<Message>
}

class Grade {
  - assignment: Assignment
  - value: Double
}

class Milestone {
  - title: String
  - date: Date
}

class Message {
  - sender: User
  - content: String
  - timestamp: Date
}

User "1" -- "1" Preferences
User "1" *-- "*" Course
Course "1" *-- "*" Assignment
Reminder "1" -- "1" User
Collaboration "1" -- "*" User
Collaboration "1" *-- "*" Discussion
Collaboration "1" *-- "*" Resource
ResourceLibrary "1" -- "*" Resource
ProgressTracking "1" -- "*" Assignment
ProgressTracking "1" -- "*" Grade
ProgressTracking "1" -- "*" Milestone
Analytics "1" -- "1" User
Analytics "1" -- "1" PerformanceData
Discussion "1" -- "*" User
Discussion "1" *-- "*" Message
@enduml
``` 

![Class diagram](https://www.plantuml.com/plantuml/png/hLJ1ZjD03BtdAw8UA_N0jKBLmXPnsPLM1Jx0JTpiY1bFP7iMhH3_pYQPp2GfBKNYL8tzwcU_P-TE55Y3iqLbGQJy9iZbpw8iDsKGP0A7s_ABig6c3xOWyiDpFGkY0sFdCCOJCbA5iYsVygFuLKIYUnzOC59L_UFUrtEoC_OQfMBJgl4qYpEA3zoJFHZHzuVutcM-3oAc8OUaaLEDsjkgrmYrDVG6teDY7n853P9GYUM0pb2D_EySeB7-L29huUWPEdIisB0FxIEw8_8uR6VPRYXYf0eYnjEOswV8W91dOAmFVvTh33sO8mE_HlORT7vYtp2AV6MelXjgaiEkjQXOPn_68ZcoYoD2dHW-TuyXxen5KKyfznW3Cvi9xAkQIY9lYtpox80gt8F2spBE0zF6qvQExZtZOl_8gpcP2dIG-AFt5e7ELRjeVs1xlZIVNf3qlsnCjdiS7rXDPLeWFLyNXoBGP3dxvrJCNlnO2P9BsyajzQaNiA5hnOUZdJIJx5cETdM4s4NyZs1tJTl-inJtY1H9PsNKE1G5r-PQ7RvSlLkLcqt_Cp4tvzPTShsAdw9Y-9bb9eEcoyvLXwvcLvgowqlPL3jxTHcK5gfOtcFcoPZbwKNGSf9hiDxnguZaQf5ERId9dpFBmylpBqLRf98OSJMAEwGwEFiR)

**Reply by ghost** [reply]
perfect 💯

**Comment by bellaabdelouahab** [comment]
fantastic 👌
i will wait for active collabs to give feedback

@yassineboujrada 
@bouslama-hamza 
@muttaqin1

**Comment by bellaabdelouahab** [comment]
but first you can update the user model with the existing one on backend
@Khalid1G

**Comment by Khalid1G** [comment]
Considering my most recent project (SPE), I believe the user class still requires a few extra fields. after including the fundamental features like RBAC and password reset. The model in use looked like this: 
```puml
class User {
  - fname: String
  - lname: String
  - password: String
  - email: String
  - password: String
  - roles: List<Message>
  - tel: String
  - delete : Boolean
  - changePasswordAt: DateTime 
  - preferences: Preferences
  - passwordResetToken: String,
  - passwordResetExpires: Date,
}
```
I think no need for an address, also about   `changePasswordAt DateTime` This field helps in JWT auth, in case of the user changes his password after generating the JWT token, the token is still valid so we need to check this case, and throw 401 error @muttaqin1.

About the fields `authorities` and  `restrictions` I don't understand the logic of it, I will add I unit test and if pass well I will merge your code and add those fields.

**Comment by bellaabdelouahab** [comment]
sure thing, those field can help with the password restore logic;
here is the issue if you want to contribute in it [#22]

we can also add other base fields at some point like "createdBy", "createdDate", "lastModifiedBy", "lastModifiedDate"

**Reply by Khalid1G** [reply]
about `createdDate` and `lastModifiedDate` is already exist, when we set `{ timestamps: true }` .

**Reply by bellaabdelouahab** [reply]
Yeah, they are my bad.

**Comment by bellaabdelouahab** [comment]
I want to add these fields 

- activationKey : when a user signup his account will not be active until he verifies his email 
- resetKey : to impl the password restore logic we need this field
- resetDate : we can also check this to prevent changing password many times 

> I need some feedbacks

@muttaqin1 @Khalid1G @yassineboujrada @bouslama-hamza

**Reply by Khalid1G** [reply]
May you need also the resetTokenExpire: the password reset token must be valid for a few minutes.
You can try to merge this field with last fields i provided

**Comment by ghost** [comment]
How many times do you plan to let the users reset their password?
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
I think each 48 hours but for now i don't know

**Reply by Khalid1G** [reply]
@bellaabdelouahab  the recommended time is about 5 to 10 min.
@muttaqin1

**Reply by ghost** [reply]
> @bellaabdelouahab  the recommended time is about 5 to 10 min.
> @muttaqin1  

Suppose i have changed my account  password.
Then my girlfriend find out my account Password and logged in.
Now i am helpless because I only can reset my password once in 48 hour's.
So I request make this time shorter if possible.
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
that's one down side of having a gf 😂.
anyway lets not add it for now,we will keep it open

**Comment by bellaabdelouahab** [comment]
about the delete field ? should i add deleted field ? and then when ever a doc deleted i just set delete to true

**Comment by bellaabdelouahab** [comment]
but there will be a problem of unique fields

**Comment by bellaabdelouahab** [comment]
@Khalid1G
